rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs

```diff
@@ -11,26 +11,135 @@
 //! [`ConfigBag`](aws_smithy_types::config_bag::ConfigBag) instead of in
 //! [`RuntimeComponents`](RuntimeComponents).
 
+use crate::box_error::BoxError;
 use crate::client::auth::{
     AuthScheme, AuthSchemeId, ResolveAuthSchemeOptions, SharedAuthScheme,
     SharedAuthSchemeOptionResolver,
 };
 use crate::client::endpoint::{ResolveEndpoint, SharedEndpointResolver};
 use crate::client::http::{HttpClient, SharedHttpClient};
 use crate::client::identity::{
-    ConfiguredIdentityResolver, ResolveIdentity, SharedIdentityResolver,
+    ConfiguredIdentityResolver, ResolveCachedIdentity, ResolveIdentity, SharedIdentityCache,
+    SharedIdentityResolver,
 };
 use crate::client::interceptors::{Intercept, SharedInterceptor};
 use crate::client::retries::classifiers::{ClassifyRetry, SharedRetryClassifier};
 use crate::client::retries::{RetryStrategy, SharedRetryStrategy};
+use crate::impl_shared_conversions;
 use crate::shared::IntoShared;
 use aws_smithy_async::rt::sleep::{AsyncSleep, SharedAsyncSleep};
 use aws_smithy_async::time::{SharedTimeSource, TimeSource};
+use aws_smithy_types::config_bag::ConfigBag;
 use std::fmt;
+use std::sync::Arc;
 
 pub(crate) static EMPTY_RUNTIME_COMPONENTS_BUILDER: RuntimeComponentsBuilder =
     RuntimeComponentsBuilder::new("empty");
 
+/// Validates client configuration.
+///
+/// This trait can be used to validate that certain required components or config values
+/// are available, and provide an error with helpful instructions if they are not.
+pub trait ValidateConfig: fmt::Debug + Send + Sync {
```

Comment:
This is great. Love the idea of having a centralized place for validating config 🎉